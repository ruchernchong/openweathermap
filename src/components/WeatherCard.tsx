import React, { Fragment } from "react";
import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Avatar,
  createTheme,
  Grid,
  Typography,
  withStyles
} from "@material-ui/core";
import { ExpandMore } from "@material-ui/icons";
import {
  formatDate,
  formatDecimal,
  formatTemperature,
  formatTime
} from "../utils";
import { TempRange, Weather } from "../types/forecast.types";

const theme = createTheme();

const Description = withStyles({
  root: {
    textTransform: "capitalize"
  }
})(Typography);

const Info = withStyles({
  root: {
    padding: theme.spacing(1, 0)
  }
})(Grid);

const WeatherIcon = withStyles({
  root: {
    width: theme.spacing(8),
    height: theme.spacing(8),
    marginRight: theme.spacing(2)
  }
})(Avatar);

interface WeatherCardProps {
  dt: number;
  humidity: number;
  sunrise?: number;
  sunset?: number;
  temp?: number;
  tempRange?: TempRange;
  weather: Weather[];
  isDailyForecast: boolean;
}

export const WeatherCard = ({
  dt,
  humidity,
  sunrise,
  sunset,
  temp,
  tempRange,
  weather,
  isDailyForecast
}: WeatherCardProps) => {
  return (
    <Accordion>
      <AccordionSummary expandIcon={<ExpandMore />}>
        <Grid container alignItems="center">
          <Grid item xs={12} md={4}>
            {weather.map(({ description, icon }, key) => (
              <Grid key={key} container alignItems="center">
                <Grid item>
                  <WeatherIcon
                    src={`http://openweathermap.org/img/wn/${icon}@2x.png`}
                    alt={`icon-${icon}`}
                  />
                </Grid>
                <Grid item>
                  <Description>{description}</Description>
                </Grid>
              </Grid>
            ))}
          </Grid>
          <Grid item xs={12} md>
            <Grid container alignItems="center">
              <Info item xs={12} md>
                {isDailyForecast
                  ? formatDate(dt)
                  : [formatDate(dt), formatTime(dt)].join(" ")}
              </Info>
              {tempRange ? (
                <Fragment>
                  <Info item xs={12} md>
                    High: {formatTemperature(tempRange.max)}
                  </Info>
                  <Info item xs={12} md>
                    Low: {formatTemperature(tempRange.min)}
                  </Info>
                </Fragment>
              ) : (
                <Info item xs={12} md>
                  Temp: {formatTemperature(temp)}
                </Info>
              )}
            </Grid>
          </Grid>
        </Grid>
      </AccordionSummary>
      <AccordionDetails>
        <Grid container>
          <Info item xs={12}>{`Humidity: ${formatDecimal(humidity, 0)}%`}</Info>
          {sunrise && sunset && (
            <Fragment>
              <Info item xs={12}>
                Sunrise: {formatTime(sunrise)}
              </Info>
              <Info item xs={12}>
                Sunset: {formatTime(sunset)}
              </Info>
            </Fragment>
          )}
        </Grid>
      </AccordionDetails>
    </Accordion>
  );
};
