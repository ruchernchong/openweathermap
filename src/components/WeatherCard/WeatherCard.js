import React, { Fragment } from "react";
import PropTypes from "prop-types";
import {
  Accordion,
  AccordionDetails,
  AccordionSummary,
  Avatar,
  createTheme,
  Grid,
  Typography,
  withStyles
} from "@material-ui/core";
import { ExpandMore } from "@material-ui/icons";

import {
  formatDate,
  formatDecimal,
  formatTemperature,
  formatTime
} from "../../utils";

const theme = createTheme();

const Description = withStyles({
  root: {
    textTransform: "capitalize"
  }
})(Typography);

const Info = withStyles({
  root: {
    padding: theme.spacing(1, 0)
  }
})(Grid);

const WeatherIcon = withStyles({
  root: {
    width: theme.spacing(8),
    height: theme.spacing(8),
    marginRight: theme.spacing(2)
  }
})(Avatar);

const WeatherCard = ({
  dt,
  humidity,
  sunrise,
  sunset,
  temp,
  tempRange,
  weather,
  isDailyForecast
}) => {
  return (
    <Accordion>
      <AccordionSummary expandIcon={<ExpandMore />}>
        <Grid container justify="center" alignItems="center">
          <Grid item xs={12} md={4}>
            {weather.map(({ description, icon, id }, key) => (
              <Grid key={key} container alignItems="center">
                <Grid item>
                  <WeatherIcon
                    src={`http://openweathermap.org/img/wn/${icon}@2x.png`}
                    alt={`icon-${icon}`}
                  />
                </Grid>
                <Grid item>
                  <Description>{description}</Description>
                </Grid>
              </Grid>
            ))}
          </Grid>
          <Grid item xs={12} md>
            <Grid container alignItems="center">
              <Info item xs={12} md>
                {isDailyForecast
                  ? formatDate(dt)
                  : [formatDate(dt), formatTime(dt)].join(" ")}
              </Info>
              {tempRange ? (
                <Fragment>
                  <Info item xs={12} md>
                    High: {formatTemperature(tempRange.max)}
                  </Info>
                  <Info item xs={12} md>
                    Low: {formatTemperature(tempRange.min)}
                  </Info>
                </Fragment>
              ) : (
                <Info item xs={12} md>
                  Temp: {formatTemperature(temp)}
                </Info>
              )}
            </Grid>
          </Grid>
        </Grid>
      </AccordionSummary>
      <AccordionDetails>
        <Grid container>
          <Info item xs={12}>{`Humidity: ${formatDecimal(humidity, 0)}%`}</Info>
          {sunrise && sunset && (
            <Fragment>
              <Info item xs={12}>
                Sunrise: {formatTime(sunrise)}
              </Info>
              <Info item xs={12}>
                Sunset: {formatTime(sunset)}
              </Info>
            </Fragment>
          )}
        </Grid>
      </AccordionDetails>
    </Accordion>
  );
};

WeatherCard.propTypes = {
  dt: PropTypes.number,
  humidity: PropTypes.number,
  sunrise: PropTypes.number,
  sunset: PropTypes.number,
  temp: PropTypes.number,
  tempRange: PropTypes.object,
  weather: PropTypes.arrayOf(PropTypes.object),
  isDailyForecast: PropTypes.bool
};

export default WeatherCard;
