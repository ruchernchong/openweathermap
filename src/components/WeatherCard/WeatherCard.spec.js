import React from "react";
import { fireEvent, render } from "@testing-library/react";
import WeatherCard from "./WeatherCard";

import mockForecast from "../../../__mocks__/forecast.json";

const mockListItem = mockForecast.data.daily[0];
const { dt, humidity, sunrise, sunset, temp, weather } = mockListItem;

describe("WeatherCard", () => {
  test("should render", () => {
    const { container } = render(
      <WeatherCard
        dt={dt}
        humidity={humidity}
        sunrise={sunrise}
        sunset={sunset}
        tempRange={temp}
        weather={weather}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should show additional contents when the accordion summary is clicked", () => {
    const { getAllByText, queryAllByText } = render(
      <WeatherCard
        dt={dt}
        humidity={humidity}
        sunrise={sunrise}
        sunset={sunset}
        tempRange={temp}
        weather={weather}
      />
    );

    expect(queryAllByText(/Humidity/)[0]).not.toBeVisible();

    fireEvent.click(getAllByText(/Tue, 25 Aug/)[0]);
    expect(getAllByText(/Humidity/)[0]).toBeVisible();
  });
});
