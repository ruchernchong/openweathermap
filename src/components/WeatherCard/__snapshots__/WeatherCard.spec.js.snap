// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCard should render 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root MuiAccordion-rounded MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-4"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
            >
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <div
                  class="MuiAvatar-root WithStyles(ForwardRef(Avatar))-root-1 MuiAvatar-circle"
                >
                  <img
                    alt="icon-10d"
                    class="MuiAvatar-img"
                    src="http://openweathermap.org/img/wn/10d@2x.png"
                  />
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item"
              >
                <p
                  class="MuiTypography-root WithStyles(ForwardRef(Typography))-root-2 MuiTypography-body1"
                >
                  light rain
                </p>
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-true"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center"
            >
              <div
                class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-true"
              >
                (Sun) 28 Jun
              </div>
              <div
                class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-true"
              >
                High: 
                30.7°C
              </div>
              <div
                class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-md-true"
              >
                Low: 
                27.7°C
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </div>
    </div>
    <div
      class="MuiCollapse-container MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container"
              >
                <div
                  class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12"
                >
                  Humidity: 66%
                </div>
                <div
                  class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12"
                >
                  Sunrise: 
                  7:01 AM
                </div>
                <div
                  class="MuiGrid-root WithStyles(ForwardRef(Grid))-root-3 MuiGrid-item MuiGrid-grid-xs-12"
                >
                  Sunset: 
                  7:13 PM
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
